ocaml/xapi/console.ml

```diff
@@ -126,7 +126,86 @@ let address_of_console __context console : address option =
     ) ;
   address_option
 
-let real_proxy' vnc_port s =
+module Console_idle_monitor = struct
+  let get_console_idle_timeout_value ~__context ~vm =
+    try
+      let idle_timeout =
+        if Db.VM.get_is_control_domain ~__context ~self:vm then
+          let host = Helpers.get_localhost ~__context in
+          Db.Host.get_console_idle_timeout ~__context ~self:host
+        else
+          let pool = Helpers.get_pool ~__context in
+          Db.Pool.get_vm_console_idle_timeout ~__context ~self:pool
+      in
+      idle_timeout
+    with _ -> 0L (* Default: no timeout *)
+
+  let has_user_activity messages =
+    List.exists
+      (function
+        | Rfb_client_msgtype_parser.KeyEvent
+        | Rfb_client_msgtype_parser.PointerEvent
+        | Rfb_client_msgtype_parser.QEMUClientMessage ->
+            true
+        | _ ->
+            false
+        )
+      messages
+
+  let is_idle_timeout ~idle_timeout_seconds ~last_activity =
+    idle_timeout_seconds > 0.
+    &&
+    let elapsed = Mtime_clock.count !last_activity in
+    Mtime.Span.to_float_ns elapsed /. 1e9 > idle_timeout_seconds
+
+  (* Create an idle timeout callback for the console,
+     if idle_timeout, then close the proxy *)
+  let create_idle_timeout_callback ~__context ~vm () =
+    let rfb_parser = ref (Some (Rfb_client_msgtype_parser.create ())) in
+    let last_activity = ref (Mtime_clock.counter ()) in
+    (* Return true for idle timeout to close the proxy, otherwise keep the proxy open *)
+    fun (buf, read_len, offset) ->
+      match !rfb_parser with
+      | None ->
+          false
+      | Some parser -> (
+        try
+          let data = Bytes.sub_string buf offset read_len in
```

Comment:
Note that this still creates a copy. I had hoped that we could avoid that and pass buf+offset+len straight to the parser. Perhaps not worth doing, but I can't remember how the parser processes this.